tvm/src/codegen/hlsc/codegen_shls.cc

```diff
@@ -0,0 +1,1359 @@
+    PrintType(p2p_dtypes[i], ss);
+    ss << "> " << p2p_names[i] << ";\n";
+  }
+
+  for (unsigned i = 0; i < offchip_mems.size(); i++) {
+    this->PrintIndentAnyStream(ss, scope);
+    PrintType(mem_dtypes[i], ss);
+    ss << " " << offchip_mems[i];
+    ss << "[";
+    int count = 0;
+    for (auto& s : mem_shapes[i]) {
+      if (count != 0) ss << "][";
+      ss << s;
+      count = count + 1;
+    }
+    ss << "];\n";
+  }
+
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "dut_wrapper *m_dut;\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "tb *m_tb;\n";
+
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "SC_CTOR(System)\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << ": clk_sig(\"clk_sig\", CLOCK_PERIOD, SC_NS)\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << ", rst_sig(\"rst_sig\")\n";
+
+  for (unsigned i = 0; i < p2p_names.size(); i++) {
+    this->PrintIndentAnyStream(ss, scope);
+    ss << ", " << p2p_names[i] << "(\"" << p2p_names[i] << "\")";
+  }
+  this->PrintIndentAnyStream(ss, scope);
+  ss << ", finish(\"finish\")\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "{\n";
+  scope += 2;
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_dut = new dut_wrapper(\"dut_wrapper\"";
+  for (unsigned i = 0; i < offchip_mems.size(); i++) {
+    ss << ", " << offchip_mems[i];
+  }
+  ss << ");\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_dut->clk(clk_sig);\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_dut->rst(rst_sig);\n";
+  for (unsigned i = 0; i < p2p_names.size(); i++) {
+    this->PrintIndentAnyStream(ss, scope);
+    ss << "m_dut->" << p2p_names[i] << "(" << p2p_names[i] << ");\n";
+  }
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_dut->finish(finish);\n\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_tb = new tb(\"tb\"";
+  for (unsigned i = 0; i < offchip_mems.size(); i++) {
+    ss << ", " << offchip_mems[i];
+  }
+  ss << ");\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_tb->clk(clk_sig);\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_tb->rst(rst_sig);\n";
+  for (unsigned i = 0; i < p2p_names.size(); i++) {
+    this->PrintIndentAnyStream(ss, scope);
+    ss << "m_tb->" << p2p_names[i] << "(" << p2p_names[i] << ");\n";
+  }
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "m_tb->finish(finish);\n\n";
+  scope -= 2;
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "}\n\n";
+
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "~System()\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "{\n";
+  scope += 2;
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "delete m_tb;\n";
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "delete m_dut;\n";
+  scope -= 2;
+  this->PrintIndentAnyStream(ss, scope);
+  ss << "}\n";
+  ss << "};\n\n";
+  ss << "#endif // _SYSTEM_H_";
+  this->support_fnames.push_back("system.h");
+  this->support_files.push_back(ss.str());
+}
+
+void CodeGenStratusHLS::GenerateTestBenchHeader(
```

Comment:
this is probably somewhere we should use a code template to simplify the codegen in the future. we do not need to use that in this PR.

Reply:
Right, I agree. The template should also allow generating multiple files for kernel design.